File: kattis/NAIPC Practice 12/B Aspen Avenue/aspen.py
from math import sqrt

n = int(input())
l, w = map(int, input().split())
points = sorted([int(input()) for _ in range(n)])
targets = []
for i in range(n // 2 + 1):
    target = i // 2 *(l / ((n - 2) // 2))
    targets.append(target)
# print(*targets)
dp = [[0 for _ in range(n // 2 + 1)] for _ in range(n // 2 + 1)]
for i in range(n // 2 + 1):
    for j in range(n // 2 + 1):
        # print(i+j, deltax[i+j], deltay[i+j])
        ans = 2 ** 31
        if i == 0 and j == 0:
            ans = 0
        if i > 0:
            ans = dp[i - 1][j] + abs(points[i + j - 1] - targets[j])
        if j > 0:
            tmp = abs(points[i + j - 1] - targets[i] + w * 1j)
            x = points[i + j - 1] - targets[i]
            tmp2 = sqrt(x * x + w * w)
            # assert (tmp == tmp2)
            ans = min(ans, dp[i][j - 1] + tmp2)
        dp[i][j] = ans
# for _ in dp:
#     print(*_, sep="\t")
print(dp[-1][-1])

File: kattis/NAIPC Practice 12/B Aspen Avenue/aspen.cpp
#include <bits/stdc++.h>

using namespace std;

double dp[1001][1001];

// double solve(int l, int r) {
//     if (dp[l][r] != 0.0) {
//         return dp[l][r];
//     }
//     if (l == 0 and r == 0) {
//         return 0.0;
//     }
//     double ans = 1.0e20;
// }

int main () {
    int n;
    cin >> n;
    double l, w;
    cin >> l >> w;
    vector<double> points;
    for (int i = 0; i < n; i ++) {
        double x;
        cin >> x;
        points.push_back(x);
    }
    
    sort(points.begin(), points.end());
    
    vector<double> targets;
    for (int i = 0; i < (n / 2) + 1; i ++) {
        double target = (double)(i / 2) * (l / ((n - 2) / 2));
        targets.push_back(target);
        // cerr << target << "\t";
    }
    // cerr << endl;
    
    for (int i = 0; i < (n / 2) + 1; i ++) {
        for (int j = 0; j < (n / 2) + 1; j ++) {
            double ans = 1.0e15;
            if (i == 0 and j == 0) {
                ans = 0;
            }
            if (i > 0) {
                ans = dp[i - 1][j] + abs(points[i + j - 1] - targets[j]);
            }
            if (j > 0) {
                double x = points[i + j - 1] - targets[i];
                // cerr << dp[i][j - 1] + sqrt(x * x + w * w) << " ";
                ans = min(ans, dp[i][j - 1] + sqrt(x * x + w * w));
            }
            dp[i][j] = ans;
            // cerr << ans << "\t";
        }
        // cerr << endl;
    }
    printf("%.10lf\n", dp[n/2][n/2]);
}
